### DISCUSSION TITLE
webRTC-policy-to-not-use-any-address.patch

### DISCUSSION THREAD

**Discussion by mehrvarz** [ungoogled-software/ungoogled-chromium-android#114] [General]
Hi. I am looking for a SystemWebview that is ungoogled and also supports WebRTC. 

Your SystemWebView seems to make use of Change-default-webRTC-policy-to-not-use-any-address.patch. Is this correct? This patch appears to prevent the creation of WebRTC ICE candidates at runtime. And without ICE candidates my WebRTC app cannot connect to any peers. So in essence it is not able to do anything. This app works fine in Ungoogled-chromium for Linux (desktop). It also works fine in Chromium, Firefox, Safari and Kiwi browser on any of the supported platforms. So it basically runs anywhere. Just not in ungoogled-chromium on Android. And also not in Bromite on Android, as they appear to use the same patch. In short: right now, in order to run my WebRTC app in a WebView, I need to use the Go variant.

I do OC understand the motivation behind the patch. But in my understanding only VPN/Tor users benefit from it. "Regular" users see no benefit. And killing WebRTC altogether is not a perfect solution. Can you share your thoughts on this? Here some (crazy?) ideas/proposals:

1: Apply the patch only to your browser, not to your SystemWebview.
2: Offer two SystemWebViews, one that supports WebRTC.
3: Provide a checkbox for the user to enable/disable the patch.
4: Let Java code disable the patch (webSettings.enableICECandidates(true))
5: Let JS code disable the patch (window.enableICECandidates(true))

**Comment by wchen342** [comment]
Normally I will say `Provide a checkbox for the user to enable/disable the patch` will be the preferred solution, however since this is WebView it has a different set of settings and UI which is (I think) partially in AOSP not in chromium code, so it is unclear whether this is doable. I am not sure what 4 means (importing webview in Java?), and 5 needs some serious hack on blink/V8 I think.

I personally believe this is a desirable feature for the users who want to use UC/Bromite so I would rather keep it somehow.

2 is the simplest but I don't have resource to compile more variants for all the apks. If you have a fast Internet connection, I will suggest simply compile yourself. The script `build.sh` has everything you need to compile an `apk` out of chromium/UC source, all you need is commenting out the patch in `patches/series` and run `./build.sh -l -a arm64 -t webview_target`.

Also which setting actually caused your app to stop working? There are three things changed: `enable_multiple_routes`, `enable_nonproxied_udp` and `enable_default_local_candidate`. If it was the last option, there is [a less aggressive patch from Vanadium that can substitute the current one](https://github.com/bromite/bromite/issues/589).

**Reply by mehrvarz** [reply]
Thank you for your response.

> which setting actually caused your app to stop working? There are three things changed: `enable_multiple_routes`, `enable_nonproxied_udp` and `enable_default_local_candidate`. If it was the last option, there is [a less aggressive patch from Vanadium that can substitute the current one](https://github.com/bromite/bromite/issues/589).

At this point I cannot answer this with certainty. But from the patch names I can imagine it could be all three. I would guess "enable_multiple_routes" is currently disabled in UC/Bromite on Android, which would explain why I don't see any ICE candidates. This needs to be enabled for WebRTC apps to work properly. (What this does: Say, you and I are in the same room and are using the same wifi network and I give you a WebRTC voice call from my phone. The app would make sure the shortest path is being used: p2p via wifi. Then while we talk on the phone you are walking away losing Wifi. The WebRTC stack would then dynamically re-route to your IP in your mobile network. This is what the ICE candidates are for.)

I need to look at what the other two patches actually do.

> 2 is the simplest but I don't have resource to compile more variants for all the apks.

2 may also be the best solution. Normally one would prefer a checkbox. But in this particular case the users are likely to fall into one of two "camps". You either use VPN/Tor and want to prevent your IP from leaking at all cost (in which case you probably prefer the save, non-checkbox variant) - or you don't and therefor might prefer the other implementation. So two separate installers may be the best in this case. Also: one can still install both webview installers and switch between them, should it be necessary.

> The script `build.sh` has everything you need to compile an `apk` out of chromium/UC source, all you need is commenting out the patch in `patches/series` and run `./build.sh -l -a arm64 -t webview_target`.

I am doing this right now. Let's see how far I get. Thank you for the info on "-t webview_target".

One quick question: Do you happen to know how I can change the name of the WebView implementation as it shows up under Developer options? I currently have UC and stock variant installed side by side and they both show up as "Android System WebView 96.0.nnnn.mm". The two icons are not the same but I am not able to tell which is which by looking at those very, very similar icons. I wish one would be called "Ungooo... System WebView" :-)

Edited to fix wrongly commented / Edited to fix typos

**Comment by mehrvarz** [comment]
I am not able to build webview. After downloading the complete src folder I run into this error:

```
/media/dave/1TB-work/ug/ungoogled-chromium-android
ERROR:root:Failed to get version info: Git command 'git log -1 --format=%H %ct --grep=^Change-Id: HEAD' in /media/dave/1TB-work/ug/ungoogled-chromium-android/src failed: rc=0, stdout='' stderr=''
Traceback (most recent call last):
  File "src/tools/download_optimization_profile.py", line 24, in <module>
    from urllib.request import urlopen
ImportError: No module named request
Removing depot_tools
Removing src
Removing ungoogled-chromium
Removing .cipd
Cloning into 'ungoogled-chromium'...
remote: Enumerating objects: 16454, done.
remote: Counting objects: 100% (921/921), done.

```
Complete build log: https://timur.mobi/li/log-2022-02-07.txt

Any idea why module urllib.request does not exist?

**Reply by wchen342** [reply]
Are you on an old version of Debian or some LTS system? The error happens when running Python 3 scripts with Python 2. The simplest fix is to get `conda` and create a Python 3-defualt environment.

**Reply by mehrvarz** [reply]
A-ha! Thank you very much for your response. I'm on Ubuntu 20-04 and I usually don't use Python. There is a Python 3.8.10 on my system and I made it the default Python now. Unfortunately your script deletes everything when an error occurs so it will take a bit until I get to the same point (50min). What if there is another issues?
And one completely unrelated question: how do I build a 32bit webview? I need both but -a arm32 does not seem to be supported.

**Reply by wchen342** [reply]
Yeah the script is made for automated builds on cloud so download speed is not a concern there. You can comment out line 102-109 so it won't delete things, but then you need to make sure if the folders already exists, `prepare_repos` is not run again otherwise `git clone` will fail.

32-bit `arm` is just `arm`, there is no 32 appendix.

**Reply by mehrvarz** [reply]
It went further this time. I see some errors fly by:

```
ERROR:root:Failed to get version info: Git command 'git log -1 --format=%H %ct --grep=^Change-Id: HEAD' in /media/dave/1TB-work/ug/ungoogled-chromium-android/src failed: rc=0, stdout='' stderr=''
ERROR:root:Failed to get version info: Git command 'git log -1 --format=%H %ct --grep=^Change-Id: HEAD' in /media/dave/1TB-work/ug/ungoogled-chromium-android/src failed: rc=0, stdout='' stderr=''
Downloading https://commondatastorage.googleapis.com/chromium-browser-clang/Linux_x64/clang-llvmorg-14-init-8033-gabb2a91b-2.tgz .......... Done.
Installing Debian sid i386 root image: /media/dave/1TB-work/ug/ungoogled-chromium-android/src/build/linux/debian_sid_i386-sysroot
...
patching file ungoogled-chromium/pruning.list
Hunk #1 succeeded at 16142 (offset -1320 lines).
ERROR: 1900 files could not be pruned.

```
It then starts patching files. Here I see very many "WARNING: Skipping non-existant path" messages. Is this normal?

Then it does some configure.ac, some checking, some elfutils, some cc and gcc. And then it aborts with this:

./build.sh: line 277: pushd: android-rebuilds: No such file or directory

Any idea what might be wrong here?

I did not modify build.sh or anything else (yet). But bc the script has aborted, no files were deleted. I am not sure I fully understand this: "but then you need to make sure if the folders already exists, prepare_repos is not run again otherwise git clone will fail." Are you suggesting I outcomment line 218 and that should be good?

**Reply by wchen342** [reply]
Yeah the warnings are normal. I forgot to mention you need android sdk/ndk rebuilds under a specific folder. They used to be automatically downloadable from a website but the original author stopped maintaining it half a year ago so I made my own fork.

So here is what you need:
1. under the same folder of `build.sh`, create `android-rebuild` folder
2. Download [sdk](https://github.com/wchen342/android-rebuilds/releases/download/sdk-12.0.0_r13/android-sdk_user-12.0.0_r13_linux-x86.zip) and [ndk](https://github.com/wchen342/android-rebuilds/releases/download/ndk-r23/android-ndk-r23-linux-x86_64.zip) and [sdk tools](https://github.com/wchen342/android-rebuilds/releases/download/sdk-repo-tools-26.1.1/sdk-repo-linux-tools-26.1.1.zip) under `android-rebuild`.
3. Start from line 244 again.

**Reply by mehrvarz** [reply]
Start from line 244, how? If I do, I run into unbound variables with no end. 40 lines before the end of the script but I have to let it rest for now. Last error: gn: command not found

**Reply by wchen342** [reply]
It is a bash script so you can just copy paste the rest of the script into console, or if it is easier then maybe just delete `src`/`ungoogled-chromium` and restart, otherwise `git` and `gclient` will run into problems and the scripts doesn't take that into account.

You need `gn` and `ninja` installed on your machine. Also `java8` and `JDK` plus a bunch of other things. The list is from Fedora, but Ubuntu shall have equivalent for most of them:

```
passwd git wget protobuf python2.7 python3 python-unversioned-command java-1.8.0-openjdk-headless java-1.8.0-openjdk-devel java-11-openjdk java-11-openjdk-headless java-11-openjdk-devel perl patch gperf rsync tar unzip bzip2 gnupg2 curl maven yasm npm ninja-build gn nodejs clang clang-tools-extra lld llvm make flex bison libdrm-devel nss-devel dbus-devel expat-devel libstdc++-static krb5-devel libatomic-static libxkbcommon-devel libuuid-devel glib2 glib2-devel glibc.i686 glibc-devel.i686 fakeroot-libs.i686 libgcc.i686 libtool-ltdl.i686 libtool-ltdl-devel.i686
```

**Reply by mehrvarz** [reply]
E: Unable to locate package gn
I have all other packages. What is gn and what format is expected for../../uc_keystore/keystore.gn?

**Reply by mehrvarz** [reply]
There's a gn script in depot_tools. But no such apt package on Ubuntu.

**Reply by wchen342** [reply]
I think this is the one for Ubuntu 20.04LTS? http://manpages.ubuntu.com/manpages/jammy/en/man1/gn.1.html.
The package name seems to be `generate-ninja`.

**Reply by wchen342** [reply]
Oh wait that is for 22.04LTS. But normally `gn` is a standalone file so you can probably use that version just fine.

**Reply by mehrvarz** [reply]
OK I was able to install generate-ninja.deb for 21.04 on 20.04 and I can now run gn. What should $TARGET_EXPANDED contain? And what format is expected in ../../uc_keystore/keystore.gn ?

**Reply by wchen342** [reply]
`keystore.gn` contains
```
android_keystore_name="your-keystore-name"
android_keystore_path="path-to-your-keystore-file"
```
, and you need a keystore file generated with Java or Android Studio, see https://stackoverflow.com/questions/11446121/how-do-i-generate-a-keystore-for-google-play.
But if you are fine with a debug key signed apk (which is less secure but if you know only you will install apks on your machine than it is fine), then you can ignore the `keystore` part.

`$TARGET_EXPANDED` is generated from `$TARGET`, see https://github.com/ungoogled-software/ungoogled-chromium-android/blob/75ae421d40201400653b84c1428bd8d39b2fd1dd/build.sh#L83.

**Reply by mehrvarz** [reply]
You tell me what ${webview_target} is. And I tell you what ${!TARGET} is. OK?

**Reply by mehrvarz** [reply]
Found out myseeeeeeeelf. I did what every sane person would have done. I checked your project out into a separate folder and ran build.sh until line 93 where it is being printed out. This way I was able to continue the script in the original folder and now it's building :-)

**Reply by mehrvarz** [reply]
I now have a Java compile issue at [21237/88969]:
```
java.lang.NoSuchMethodError: java.nio.ByteBuffer.position(I)Ljava/nio/ByteBuffer
  at com.google.turbine.zip.Zip$ZipIterator.string(Zip.java:139)
```
Does this ring a bell? I am using Java8. Complete log: https://timur.mobi/li/log-2022-02-08.txt

**Reply by wchen342** [reply]
The compilation needs to use Java 10+ as default, Ubuntu can set default with `update-java-alternatives --list` and then `sudo update-java-alternatives --set /path/to/java/version` I believe.
`

**Reply by mehrvarz** [reply]
OK. I run ninja from the command line now like so: ninja -C "out/Default" "system_webview_apk"
After switching to Java11, how can I clean up the generated Java8 bytecode and start from scratch?

**Reply by wchen342** [reply]
It shall be `ninja -C out/Default -t clean`.

**Reply by mehrvarz** [reply]
After switching to Java11 and doing "ninja -t clean" and starting ninja again I can see that it is still using "-source 1.8 -target 1.8" to build. Where could this be coming from? A config file?

Also this: java.io.IOException: error reading ../../third_party/jdk/extras/java_8/jre/lib/rt.jar
I think to myself, it shouldn't use Java8, but why is lib/rt.jar not found? Turns out third_party/jdk/extras/java_8/jre is a symb-link to /usr/lib/jvm/jre-1.8.0. But this does not exist on Ubuntu. On Ubuntu jre1.8 is in that folder but is callled differently. Could this symb-link be something specific to your Linux dist?

**Reply by wchen342** [reply]
Yes I think you need to adjust the paths for all the `ln -s` commands. The paths are for Fedora and you need to find them on Ubuntu and give the right paths to them.

I'm not sure about the "-source 1.8 -target 1.8" thing, there are certain java8 files used, that's why java8 is still needed. As long as there is no error then it shall be fine.

**Reply by mehrvarz** [reply]
I went to /usr/lib/jvm and did this: ln -s java-8-openjdk-amd64/jre jre-1.8.0
This does indeed help me over the IOException and the build is running again. Let's see how far I get...

**Comment by mehrvarz** [comment]
Got it running on the device! Re-activated WebRTC (it's working!) and changed the apk-label in android_webview/nonembedded/java/AndroidManifest.xml 
Check this out: https://timur.mobi/li/Screenshot_20220209-063600_Settings_.png
Also verified that it does not phone home :-)
Thank you so much for your help. If you were to create a build recipe page/document, I would be happy to contribite (with my Ubuntu angle at least). Such a recipe doesn't need to be 100% perfect (which would be a lot of work), just a jump starter for the next guy. Again, thank you for helping me!
Please consider adding these very minimal build script changes:

```
--- a/build.sh
+++ b/build.sh
-mkdir "ndk_temp"
+mkdir -p "ndk_temp"
-mkdir android-sdk
-mkdir android-ndk
+mkdir -p android-sdk
+mkdir -p android-ndk
```

BTW, I think it would be more correct for you to say (under "Differences from ungoogled-chromium") "Disables WebRTC to prevent address leaking" instead of the half truth "Prevent WebRTC address leaking" which sounds as if WebRTC was still working.

**Reply by wchen342** [reply]
Glad to see it worked! If you want to write a tutorial you can paste it here and I will copy it to wiki.

> "Disables WebRTC to prevent address leaking" instead of the half truth "Prevent WebRTC address leaking"

It does work for, e.g. meetings, so it is not completely disabled. It just happens to be the case that your use case is in conflict with what we use to prevent address leak.

**Reply by mehrvarz** [reply]
Do you know how I can have two webview variants installed side by side? The package names need to be different. This is all. It works with the G variant because they use com.google.android.webview. To have two ungoogle variants installed side by side (my proposal 2 from post 1), one of the two has to change the package name. I have tried changing it in the manifest file, where I changed the apk name, but this isn't good enough. When I try to install the compiled apk, Android says that this package is already installed. Changing the package name in the manifest seems to do nothing.

**Reply by wchen342** [reply]
I remember even if you can change the webview package name it will not work, because in Android webview are limited to certain names.
@csagan5 I saw you updated the Bromite wiki page about webview and now you have a custom Webview name `org.bromite.webview`, how is that loadable?

**Reply by mehrvarz** [reply]
On https://github.com/bromite/bromite/wiki/Installing-SystemWebView it says "You can install Bromite SystemWebView only if the above output (aapt d xmltree framework-res.apk...) contains a packagename="org.bromite.webview" with the correct signature block or without a corresponding E: signature block." Since none of my devices does this (I just checked), how could I ever install and use a bromite webview?

Also: "NOTE: without signature it will be UNSECURE" - What does this mean? What type of danger do you have in mind? Some 3rd party app being able to replace the complete webview without me realizing?

Anyways, according to this wiki the idea of letting a user switch between two 3rd party webviews appears to be not doable. In the best case one can switch btw one (or more) G webview(s) and max one 3rd party webview. I am sad now.

**Reply by wchen342** [reply]
Yeah if my understanding is correct, for that to work you either need a Debug build of Android, a custom build or rooted system with a Magisk module that can change some system files. Otherwise you can only use fixed names for webview. It is a limitation in Android system unfortunately so it cannot be fixed from here.

**Reply by csagan5** [reply]
> @csagan5 I saw you updated the Bromite wiki page about webview and now you have a custom Webview name `org.bromite.webview`, how is that loadable?

Same as it was before; until now you needed either a pre-installed Bromite SystemWebView in the ROM and no signature pinned for `com.android.webview` or the Bromite signature pinned for it. The package rename does not cut out any hack or specific use case, except the dangerous and unsecure ones where a ROM would ship without webview pre-installed and without signatures pinned (there are none anyways, as far as I know).

> Also: "NOTE: without signature it will be UNSECURE" - What does this mean? What type of danger do you have in mind? Some 3rd party app being able to replace the complete webview without me realizing?

If your ROM has no preinstalled webview and no pinned signature, but the package is on the allow list, then it is a huge security vulnerability. It happened at least once for a popular ROM, but nowadays nobody will repeat this mistake (hopefully).

**Reply by wchen342** [reply]
I see, so the signature pinning is still needed. However I don't know any ROM that preload with Bromite (or maybe I missed something there?), how do you get your WebView pre-installed or get your signature pre-loaded in the ROM without root?

**Reply by csagan5** [reply]
It is not possible without root and never has been.
